File: java/src/main/java/fr/sdecout/handson/persistence/author/DbAuthorAdapter.java
package fr.sdecout.handson.persistence.author;

import fr.sdecout.handson.rest.author.*;
import jakarta.transaction.Transactional;
import org.springframework.stereotype.Component;

import java.util.Optional;
import java.util.stream.Stream;

@Component
@Transactional
class DbAuthorAdapter implements AuthorAccess, AuthorCreation, AuthorSearch {

    private final AuthorRepository authorRepository;

    DbAuthorAdapter(AuthorRepository authorRepository) {
        this.authorRepository = authorRepository;
    }

    @Override
    public Optional<AuthorResponse> findAuthor(AuthorId id) {
        return authorRepository.findById(id.value())
                .map(AuthorEntity::toAuthorResponse);
    }

    @Override
    public Stream<AuthorSearchResponseItem> searchAuthors(String hint) {
        return authorRepository.findByLastNameLikeIgnoringCase('%' + hint + '%')
                .map( AuthorEntity::toAuthorSearchResponseItem);
    }

    @Override
    public AuthorId addAuthor(String lastName, String firstName) {
        var authorId = AuthorId.next();
        authorRepository.save(AuthorEntity.from(authorId.value(), lastName, firstName));
        return authorId;
    }

}

File: kotlin/src/main/kotlin/fr/sdecout/handson/persistence/author/DbAuthorAdapter.kt
package fr.sdecout.handson.persistence.author

import fr.sdecout.handson.rest.author.*
import jakarta.transaction.Transactional
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Component

@Component
@Transactional
class DbAuthorAdapter(
    private val authorRepository: AuthorRepository,
) : AuthorAccess, AuthorCreation, AuthorSearch {

    override fun findAuthor(id: AuthorId): AuthorResponse? = authorRepository.findByIdOrNull(id.value)
        ?.toAuthorResponse()

    override fun searchAuthors(hint: String): List<AuthorSearchResponseItem> = authorRepository
        .findByLastNameLikeIgnoringCase("%$hint%")
        .map { it.toAuthorSearchResponseItem() }

    override fun addAuthor(lastName: String, firstName: String): AuthorId = AuthorId.next().also {
        authorRepository.save(
            AuthorEntity(
                id = AuthorId.next().value,
                lastName = lastName,
                firstName = firstName,
                publishedBooks = mutableListOf()
            )
        )
    }

}
